src/gwt/src/org/rstudio/core/client/PriorityTaskQueue.java

```diff
@@ -0,0 +1,143 @@
+/*
+ * PriorityTaskQueue.java
+ *
+ * Copyright (C) 2020 by RStudio, PBC
+ *
+ * Unless you have received this program directly from RStudio pursuant
+ * to the terms of a commercial license agreement with RStudio, then
+ * this program is licensed to you under the terms of version 3 of the
+ * GNU Affero General Public License. This program is distributed WITHOUT
+ * ANY EXPRESS OR IMPLIED WARRANTY, INCLUDING THOSE OF NON-INFRINGEMENT,
+ * MERCHANTABILITY OR FITNESS FOR A PARTICULAR PURPOSE. Please refer to the
+ * AGPL (http://www.gnu.org/licenses/agpl-3.0.txt) for more details.
+ *
+ */
+
+
+package org.rstudio.core.client;
+
+import java.util.LinkedList;
+import java.util.Queue;
+
+import com.google.gwt.user.client.Command;
+
+// Task queue that allows tasks to bump themselves up in priority (even after
+// they have been added to the queue).
+
+public class PriorityTaskQueue
+{
+   public interface Task
+   {
+      String getLabel(); // used for debug/log output 
+      boolean hasPriority();
+      void execute(Command done);
+   }
+   
+   public PriorityTaskQueue()
+   {
+      this(true, false);
+   }
+   
+   public PriorityTaskQueue(boolean safe)
+   {
+      this(safe, false);
+   }
+   
+   public PriorityTaskQueue(boolean safe, boolean log)
+   {
+      log_ = log;
+      safe_ = safe;
+   }
+   
+   public void addTask(Task task)
+   {
+      log("adding " + task.getLabel());
+      taskQueue_.add(task);
+      processQueue();
+   }
+   
+   private void processQueue()
+   {
+      if (processing_)
+      {
+         log("already running");
+         return;
+      }
+      
+      processing_ = true;
+      processNextTask();
+   }
+   
+   private void processNextTask()
+   {
+      log("process next task");
+      
+      if (taskQueue_.isEmpty())
+      {
+         log("done");
+         processing_ = false;
+         return;
+      }
+      
+      // see if any of the tasks have priority
+      Task nextTask = null;
+      for (Task task : taskQueue_)
+      {
+         if (task.hasPriority())
+         {
+            nextTask = task;
+            log("executing " + nextTask.getLabel() + " [Priority]");
+            break;
+         }
+      }
+      
+      // if there is no priority task then just remove from the queue
+      if (nextTask == null)
+      {
+         nextTask = taskQueue_.peek();
+         log("executing " + nextTask.getLabel());  
+      }
+      
+      // remove the task
+      taskQueue_.remove(nextTask);
+      
+      // run the next task and then continue processing. catch any exceptions
+      // so that we can continue processing if 'safe' was requested
+      try
+      {
+         nextTask.execute(() -> {
+            log("continuation");
+            processNextTask(); 
```

Comment:
If the intent here (and elsewhere) is to handle a lot of expensive work, should we defer the processing of the next task so that the browser event loop has a chance to run between tasks?

Reply:
In this case I happened to "know" that the task would be making an RPC call (so we would be pumping events) but in general this of course won't necessarily be the case. Added a call to `scheduledDeferred()`.